### PR TITLE
cos satellite bucket creation

### PR DESCRIPTION
**Pull request by Priyaranimishra**
Adding support for creation of bucket on Satellite location using existing cos instance.

Here is the test result of each function:-
```
2022/04/18 07:02:05 Create endpointttt:.... s3.756ad6b6-72a6-4e55-8c94-b02e51e708b3.c8jh7hfw0ppoapdqrmpg.cloud-object-storage.appdomain.cloud
2022/04/18 07:02:05 Deleting the bucketss... {
  Bucket: "terraform94"
}
2022/04/18 07:02:05 Deleted the bucket:
--- PASS: TestAccIBMCosBucket_Basic_satellite (9.44s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/cos	9.469s
```

```
2022/04/18 07:56:56 Deleting the bucketss... {
  Bucket: "terraform85"
}
2022/04/18 07:56:56 Deleted the bucket:
--- PASS: TestAccIBMCosBucket_Satellite_Expiredays (21.26s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/cos	21.298s
```
```
  Bucket: "terraform21"
}
2022/04/18 08:06:53 Deleted the bucket:
--- PASS: TestAccIBMCosBucket_Satellite_Object_Versioning (7.53s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/cos	7.556s


```

```
2022/04/18 08:33:40 Deleting the bucketss... {
  Bucket: "terraform71"
}
2022/04/18 08:33:40 Deleted the bucket:
--- PASS: TestAccIBMCosBucket_Satellite_Expiredate (22.71s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/cos	22.744s
```
```

2022/04/18 09:22:22 Deleting the bucketss... {
  Bucket: "terraform33"
}
2022/04/18 09:22:22 Deleted the bucket:
--- PASS: TestAccIBMCosBucket_Satellite_Expireddeletemarker (10.59s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/cos	10.618s
```

```
2022/04/18 09:24:19 Deleting the bucketss... {
  Bucket: "terraform30"
}
2022/04/18 09:24:19 Deleted the bucket:
--- PASS: TestAccIBMCosBucket_Satellite_AbortIncompeleteMPU (13.92s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/cos	13.948s
```

```
2022/04/18 09:45:14 Deleting the bucketss... {
  Bucket: "terraform62"
}
2022/04/18 09:45:14 Deleted the bucket:
--- PASS: TestAccIBMCosBucket_Satellite_noncurrentversion (14.54s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/cos	14.572s
```

### CODE REVIEW

File: ibm/service/cos/data_source_ibm_cos_bucket.go [hunk 1]
```diff
@@ -327,14 +331,25 @@ func dataSourceIBMCosBucketRead(d *schema.ResourceData, meta interface{}) error
 	serviceID := d.Get("resource_instance_id").(string)
 	bucketType := d.Get("bucket_type").(string)
 	bucketRegion := d.Get("bucket_region").(string)
-	var endpointType = d.Get("endpoint_type").(string)
+	endpointType := d.Get("endpoint_type").(string)
+	stlocation := d.Get("satellite_location_id").(string)
+
+	if bucketType == "satellite_location_id" {
```
Comment (hkantare) [inline ibm/service/cos/data_source_ibm_cos_bucket.go:337]: bucket_type user will give has "satellite_location_id"? I think this is wrong if user provides "satellite_location_id" then you need to compute  bucket type as sl

File: ibm/service/cos/data_source_ibm_cos_bucket.go [hunk 2]
```diff
@@ -420,6 +442,11 @@ func dataSourceIBMCosBucketRead(d *schema.ResourceData, meta interface{}) error
 	if endpointType == "private" {
 		sess.SetServiceURL("https://config.private.cloud-object-storage.cloud.ibm.com/v1")
 	}
+
+	satconfig := fmt.Sprintf("https://config.%s.%s.cloud-object-storage.appdomain.cloud/v1", serviceID, bucketType)
```
Comment (hkantare) [inline ibm/service/cos/data_source_ibm_cos_bucket.go:446]: these changes should be done only if it satellite bucket

File: ibm/service/cos/resource_ibm_cos_bucket.go [hunk 3]
```diff
@@ -649,7 +666,19 @@ func resourceIBMCOSBucketUpdate(d *schema.ResourceData, meta interface{}) error
 	bucketName := parseBucketId(d.Id(), "bucketName")
 	serviceID := parseBucketId(d.Id(), "serviceID")
 	endpointType := parseBucketId(d.Id(), "endpointType")
+	bLocation := parseBucketId(d.Id(), "bLocation")
+	apiType := parseBucketId(d.Id(), "apiType")
+
+	if apiType == "sl" {
+		fmt.Println("Upate:-The substring is not present in the string.")
```
Comment (hkantare) [inline ibm/service/cos/resource_ibm_cos_bucket.go:673]: remove this fmt

File: ibm/service/cos/resource_ibm_cos_bucket.go [hunk 4]
```diff
@@ -820,6 +850,10 @@ func resourceIBMCOSBucketUpdate(d *schema.ResourceData, meta interface{}) error
 		sess.SetServiceURL("https://config.private.cloud-object-storage.cloud.ibm.com/v1")
 	}
 
+	// Add Sat Config changess
```
Comment (hkantare) [inline ibm/service/cos/resource_ibm_cos_bucket.go:853]: these changes shd be only if its satellite one right?

File: ibm/service/cos/resource_ibm_cos_bucket.go [hunk 5]
```diff
@@ -900,7 +934,7 @@ func resourceIBMCOSBucketUpdate(d *schema.ResourceData, meta interface{}) error
 		hasChanged = true
 		updateBucketConfigOptions.MetricsMonitoring = metricsMonitor
 	}
-
+	log.Println("updateBucketConfigOptions:;", updateBucketConfigOptions)
```
Comment (hkantare) [inline ibm/service/cos/resource_ibm_cos_bucket.go:937]: remove this

File: ibm/service/cos/resource_ibm_cos_bucket.go [hunk 6]
```diff
@@ -1010,6 +1062,10 @@ func resourceIBMCOSBucketRead(d *schema.ResourceData, meta interface{}) error {
 		sess.SetServiceURL("https://config.private.cloud-object-storage.cloud.ibm.com/v1")
 	}
 
+	satconfig := fmt.Sprintf("https://config.%s.%s.cloud-object-storage.appdomain.cloud/v1", serviceID, bLocation)
+
```
Comment (hkantare) [inline ibm/service/cos/resource_ibm_cos_bucket.go:1066]: same as above

File: ibm/service/cos/resource_ibm_cos_bucket.go [hunk 7]
```diff
@@ -1176,7 +1261,9 @@ func resourceIBMCOSBucketCreate(d *schema.ResourceData, meta interface{}) error
 	s3Client := s3.New(s3Sess, s3Conf)
 
 	_, err = s3Client.CreateBucket(create)
+
 	if err != nil {
+		fmt.Println("error while creating buckets", err)
```
Comment (hkantare) [inline ibm/service/cos/resource_ibm_cos_bucket.go:1266]: remove this

File: ibm/service/cos/resource_ibm_cos_bucket.go [hunk 8]
```diff
@@ -1248,8 +1345,11 @@ func resourceIBMCOSBucketDelete(d *schema.ResourceData, meta interface{}) error
 	delete := &s3.DeleteBucketInput{
 		Bucket: aws.String(bucketName),
 	}
+	log.Println("Deleting the bucketss...", delete)
```
Comment (hkantare) [inline ibm/service/cos/resource_ibm_cos_bucket.go:1348]: delete this log

File: ibm/service/cos/resource_ibm_cos_bucket.go [hunk 9]
```diff
@@ -1248,8 +1345,11 @@ func resourceIBMCOSBucketDelete(d *schema.ResourceData, meta interface{}) error
...
 	_, err = s3Client.DeleteBucket(delete)
 
+	log.Println("Deleted the bucket:")
```
Comment (hkantare) [inline ibm/service/cos/resource_ibm_cos_bucket.go:1351]: same as above

File: website/docs/d/cos_bucket.html.markdown [hunk 10]
```diff
@@ -1,11 +1,12 @@
 ---
-subcategory: "Object Storage"
+subcategory: "Object Storage" , "COS Satellite"
```
Comment (hkantare) [inline website/docs/d/cos_bucket.html.markdown:2]: Don't change the sub category

File: website/docs/d/cos_bucket.html.markdown [hunk 11]
```diff
@@ -1,11 +1,12 @@
...
+page_title: "IBM : Cloud Object Storage Bucket" , "COS Satellite Bucket"
 description: |-
-  Get information about IBM Cloud Object Storage bucket.
+  Get information about COS Satellite bucket. , "Get information about COS Satellite bucket."
 ---
 
+
 # ibm_cos_bucket
 
 Creates an IBM Cloud Object Storage bucket. It also allows object storage buckets to be updated and deleted. The ibmcloud_api_key used by Terraform must have been granted sufficient IAM rights to create and modify IBM Cloud Object Storage buckets and have access to the Resource Group the Cloud Object Storage bucket will be associated with. See https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-iam for more details on setting IAM and Access Group rights to manage COS buckets.
```
Comment (hkantare) [inline website/docs/d/cos_bucket.html.markdown:12]: Edit this description to say this resource works for IBM cloud COS and Satellite based COS

File: website/docs/d/cos_bucket.html.markdown [hunk 12]
```diff
@@ -35,6 +36,31 @@ output "bucket_private_endpoint" {
...
 ```
 
+# cos satellite bucket
+
+Create a cos satellite bucket. See the architecture https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-about-cos-satellite for more details. We are using existing cos instance to create bucket , so no need to create any cos instance via a terraform. Cos satellite does not support all features see the section **What features are currently supported?** in https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-about-cos-satellite.
+IBM Satellite documentation https://cloud.ibm.com/docs/satellite?topic=satellite-getting-started . We are supporting object versioning and expiration features as of now. Firewall is not supported yet.
+
+## Example usage
+
+```terraform
```
Comment (hkantare) [inline website/docs/d/cos_bucket.html.markdown:46]: This doc is for datasource please provide a valid example to read a bucket for satellite based COS

File: website/docs/d/cos_bucket.html.markdown [hunk 13]
```diff
@@ -44,6 +70,7 @@ Review the argument references that you can specify for your data source.
 - `endpoint_type` - (Optional, String) The type of the endpoint either `public` or `private` or `direct` to be used for the buckets. Default value is `public`.
 - `resource_instance_id` - (Required, String) The ID of the IBM Cloud Object Storage service instance for which you want to create a bucket.
 - `storage_class`- (Required, String)  Storage class of the bucket. Supported values are `standard`, `vault`, `cold`, `smart`.
+- `satellite_location_id` - (Required, String) satellite location id. Provided by end users.
```
Comment (hkantare) [inline website/docs/d/cos_bucket.html.markdown:73]: its optional and make other fields like bucket_type,...son as optioanl as per new changes

File: ibm/service/cos/resource_ibm_cos_bucket.go [hunk 14]
```diff
@@ -1121,27 +1209,53 @@ func resourceIBMCOSBucketCreate(d *schema.ResourceData, meta interface{}) error
...
 		apiType = "ssl"
 	}
+	//Add satellite location id
+	if bucketLocation, ok := d.GetOk("satellite_location_id"); ok {
+		bLocation = bucketLocation.(string)
+		apiType = "sl"
+	}
+
 	if bLocation == "" {
 		return fmt.Errorf("Provide either `cross_region_location` or `region_location` or `single_site_location`")
```
Comment (hkantare) [inline ibm/service/cos/resource_ibm_cos_bucket.go:1219]: Can you add here satellite_location_id

File: ibm/service/cos/resource_ibm_cos_bucket.go [hunk 15]
```diff
@@ -1293,14 +1428,36 @@ func resourceIBMCOSBucketExists(d *schema.ResourceData, meta interface{}) (bool,
...
+		if endpointType == "private" {
+			apiEndpoint = apiEndpointPrivate
+		}
+		if endpointType == "direct" {
+			apiEndpoint = directApiEndpoint
+		}
+
 	}
+
+	apiEndpoint = conns.EnvFallBack([]string{"IBMCLOUD_COS_ENDPOINT"}, apiEndpoint)
```
Comment (hkantare) [inline ibm/service/cos/resource_ibm_cos_bucket.go:1459]: its repeated twice

File: examples/ibm-cos-bucket/variables.tf [hunk 16]
```diff
@@ -43,7 +43,7 @@ variable "expire_days" {
 }
 
 variable "expire_date" {
-  default = ""
+  default = "2022-06-09"
```
[the remark this answers is not in the document]
Comment (Priyaranimishra) [reply examples/ibm-cos-bucket/variables.tf:46]: yes Harini. Eeven internal customer need format one. SO I have added.

**Pull request merged**